myceli/src/listener.rs

```diff
@@ -0,0 +1,122 @@
+use anyhow::Result;
+use local_storage::provider::SqliteStorageProvider;
+use local_storage::storage::Storage;
+use std::net::SocketAddr;
+use std::path::PathBuf;
+use std::rc::Rc;
+use std::time::Duration;
+use tokio::net::UdpSocket;
+use tokio::time::sleep;
+use tracing::{debug, error, info};
+
+use crate::handlers;
+use crate::receiver::Receiver;
+use messages::{ApplicationAPI, Message, MessageChunker, SimpleChunker};
+
+// TODO: Make this configurable
+pub const MTU: u16 = 60; // 60 for radio
+
+pub struct Listener {
+    storage: Rc<Storage>,
+    sender_addr: Option<SocketAddr>,
+    chunker: SimpleChunker,
+    receiver: Receiver,
+    socket: Rc<UdpSocket>,
+}
+
+impl Listener {
+    pub async fn new(listen_address: &str) -> Result<Self> {
+        let provider = SqliteStorageProvider::new("storage.db")?;
+        provider.setup()?;
+        let storage = Rc::new(Storage::new(Box::new(provider)));
+        let socket = UdpSocket::bind(&listen_address).await?;
+        info!("Listening for messages on {}", listen_address);
+        let receiver = Receiver::new(Rc::clone(&storage));
+        Ok(Listener {
+            storage,
+            sender_addr: None,
+            chunker: SimpleChunker::new(MTU),
+            receiver,
+            socket: Rc::new(socket),
+        })
+    }
+
+    pub async fn listen(&mut self) -> Result<()> {
+        let mut buf = vec![0; usize::from(MTU)];
+        loop {
+            {
+                loop {
+                    if let Ok((len, sender)) = self.socket.try_recv_from(&mut buf) {
+                        if len > 0 {
+                            self.sender_addr = Some(sender);
+                            break;
+                        }
+                    }
+                    sleep(Duration::from_millis(10)).await;
+                }
+            }
+
+            match self.chunker.unchunk(&buf) {
+                Ok(Some(msg)) => match self.handle_message(msg).await {
+                    Ok(Some(resp)) => self.transmit_response(resp).await.unwrap(),
+                    Ok(None) => {}
+                    Err(e) => error!("{e}"),
+                },
+                Ok(None) => {
+                    debug!("No msg found yet");
+                }
+                Err(err) => {
+                    error!("Message parse failed: {err}");
+                }
+            }
+        }
+    }
+
+    async fn handle_message(&mut self, message: Message) -> Result<Option<Message>> {
+        let resp = match message {
```

Comment:
This new "passing responses back from the message handlers" logic is a nice cleanup, but I'm not completely satisfied with it yet. Some messages will never have a response that is handled this way, so I may eventually split this match into one with handlers with responses-to-be-sent, and one with handlers without those responses.